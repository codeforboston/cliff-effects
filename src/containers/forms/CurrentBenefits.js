import { connect } from 'react-redux';

import { CurrentBenefitsStep } from '../../forms/CurrentBenefits';
import { setHasBenefit } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),

    USState: state.getIn([
      'geography',
      'state',
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setHasBenefit({ benefit, value }) {
      dispatch(
        setHasBenefit({
          time: 'current',
          benefit,
          value,
        })
      );
    },
  };
};

const CurrentBenefitsContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentBenefitsStep);

export { CurrentBenefitsContainer as CurrentBenefitsStep };
