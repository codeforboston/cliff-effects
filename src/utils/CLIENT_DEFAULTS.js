import { cloneDeep } from 'lodash';

/**
 * Default client values.
 * MUST ALWAYS BE VALID
 * 
 * Note: A 'positive number' money value can have more than
 *     two decimal places.
 */
const CLIENT_DEFAULTS = {

  // Property Name                | Default Value   | Valid Value Description

  current: {
    // Current programs
    hasSnap:                        false,            // boolean
    hasHousing:                     false,            // boolean
    // Household
    // 'm_' for 'member'
    household:  [{                  // 'head' member  // one or more member objects
        m_age:                      30,               // whole number
        m_role:                     'head',           // 'head', 'spouse', 'member'
        m_disabled:                 false             // boolean
    }],
    // Income
    earned:                         0,                // positive number (can have more than two decimal places)
    TAFDC:                          0,                // positive number
    SSI:                            0,                // positive number
    SSDI:                           0,                // positive number
    childSupportIn:                 0,                // positive number
    unemployment:                   0,                // positive number
    workersComp:                    0,                // positive number
    pension:                        0,                // positive number
    socialSecurity:                 0,                // positive number
    alimony:                        0,                // positive number
    otherIncome:                    0,                // positive number
    incomeExclusions:               0,                // positive number
    // Expenses
    childDirectCare:                0,                // positive number
    childBeforeAndAfterSchoolCare:  0,                // positive number
    childTransportation:            0,                // positive number
    childOtherCare:                 0,                // positive number
    earnedBecauseOfChildCare:       0,                // positive number
    childSupportPaidOut:            0,                // positive number
    adultDirectCare:                0,                // positive number
    adultTransportation:            0,                // positive number
    adultOtherCare:                 0,                // positive number
    disabledAssistance:             0,                // positive number
    earnedBecauseOfAdultCare:       0,                // positive number
    disabledMedical:                0,                // positive number
    otherMedical:                   0,                // positive number
    /** @todo When client has section 8, switch this to 'housingVoucher' */
    /** @todo Make shelter expense values more robust. */
    shelter:                        'homeless',       // 'homeless', 'homeowner', 'renter', 'housingVoucher'
    contractRent:                   0,                // positive number
    rentShare:                      0,                // positive number
    rent:                           0,                // positive number
    mortgage:                       0,                // positive number
    housingInsurance:               0,                // positive number
    propertyTax:                    0,                // positive number
    climateControl:                 false,            // boolean
    nonHeatElectricity:             false,            // boolean
    phone:                          false,            // boolean
    fuelAssistance:                 false,            // boolean  Note: Input is 'yes', 'no', but should be being converted
    otherExpenses:                  0                 // positive number
  },

  // Note: A `future` value will default to the client's `current` value until a user changes it explicitly
  future: {},                       // same structure // same

};

CLIENT_DEFAULTS.future = cloneDeep( CLIENT_DEFAULTS.current );


export { CLIENT_DEFAULTS };
