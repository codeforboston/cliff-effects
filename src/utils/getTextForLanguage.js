/** Returns a translator based on the language given */

// TRANSFORMER FUNCTIONS
import { mergeWith } from 'lodash';
import { interpolateSnippets } from './interpolation.js';

// DATA
import { getLocalizationData } from '../localization/all';
import inlineComponents from '../localization/inlineComponents';

// Get copy of localization data
const localizations = getLocalizationData();

// store interpolated and (if necessary) merged snippets objects
let finishedSnippets = { en: interpolateSnippets(localizations.en, inlineComponents) };

/** Customizes Lodash's mergeWith function to replace arrays completely
 * (to avoid arrays of English strings being mixed with arrays of translated
 * strings, if they happen to have different lengths).
 */
const mergeCustomizer = function (objValue, srcValue) {
  if (Array.isArray(objValue)) {
    return srcValue;
  }
};

/** Returns the object named by langName that contains
 * the text snippets of that language. If that language
 * doesn't exist, it warns the coder and returns English.
 */
const getTextForLanguage = function (langCode) {
  if (!localizations[ langCode ]) {
    console.warn('There\'s no localization for ' + langCode + '. Defaulting to English.');
    langCode = 'en';
  }

  if (!finishedSnippets[ langCode ]) {
    // interpolate snippets and merge with English (filling in any gaps)
    const interpolatedSnippets = interpolateSnippets(localizations[ langCode ], inlineComponents);

    // deeply merge the object containing snippets in langName with English,
    // so that we fall back to English if a particular field is missing.
    finishedSnippets[ langCode ] =  mergeWith({}, finishedSnippets.en, interpolatedSnippets, mergeCustomizer);
  }

  return finishedSnippets[ langCode ];
};  // End getTextForLanguage()


export { getTextForLanguage };
