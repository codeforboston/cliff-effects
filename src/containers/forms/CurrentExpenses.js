import { connect } from 'react-redux';

import { CurrentExpensesStep } from '../../forms/CurrentExpenses';

import { setCashValue, setHousingType, setPaysUtility, setGetsFuelAssistance } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setExpenseValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },

    setHousingType({ housingType }) {
      dispatch(
        setHousingType({
          time: 'current',
          housingType,
        })
      );
    },

    setPaysUtility({ utility, paysUtility }) {
      dispatch(
        setPaysUtility({
          time: 'current',
          utility,
          paysUtility,
        })
      );
    },

    setGetsFuelAssistance({ getsAssistance }) {
      dispatch(
        setGetsFuelAssistance({
          time: 'current',
          getsAssistance,
        })
      );
    },
  };
};

const CurrentExpensesStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentExpensesStep);

export { CurrentExpensesStepContainer as CurrentExpensesStep };
