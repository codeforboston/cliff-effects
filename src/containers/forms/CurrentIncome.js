import { connect } from 'react-redux';

import { CurrentIncomeStep } from '../../forms/CurrentIncome';

import { setCashValue } from '../../actions';

const mapStateToProps = (state) => {
  return {
    currentClient: state.getIn([
      'client',
      'current', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setIncomeValue({ name, value }) {
      dispatch(
        setCashValue({
          time: 'current',
          name,
          value,
        })
      );
    },
  };
};

const CurrentIncomeStepContainer = connect(mapStateToProps, mapDispatchToProps)(CurrentIncomeStep);

export { CurrentIncomeStepContainer as CurrentIncomeStep };
