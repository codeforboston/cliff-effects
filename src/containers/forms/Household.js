import { connect } from 'react-redux';

import { HouseholdStep } from '../../forms/Household';
import {
  addMember,
  removeMember,
  setMemberIsDisabled,
  setMemberRole,
  setMemberAge,
} from '../../actions';

const time = 'current';

const mapStateToProps = (state) => {
  return {
    household: state.getIn([
      'client',
      'current',
      'household', 
    ]),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setMemberAge({ index, age }) {
      dispatch(
        setMemberAge({
          time,
          index,
          age,
        })
      );
    },

    setMemberIsDisabled({ index, isDisabled }) {
      dispatch(
        setMemberIsDisabled({
          time,
          index,
          isDisabled,
        })
      );
    },

    setMemberRole({ index, role }) {
      dispatch(
        setMemberRole({
          time,
          index,
          role,
        })
      );
    },

    removeMember({ index }) {
      dispatch(
        removeMember({
          time,
          index,
        })
      );
    },

    addMember({ member }) {
      dispatch(
        addMember({
          time,
          member,
        })
      );
    },
  };
};

const HouseholdStepContainer = connect(mapStateToProps, mapDispatchToProps)(HouseholdStep);

export { HouseholdStepContainer as HouseholdStep };
