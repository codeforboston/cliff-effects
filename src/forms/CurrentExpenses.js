// REACT COMPONENTS
import React from 'react';
import {
  Form,
  Radio,
  Checkbox,
  Header,
} from 'semantic-ui-react';

// PROJECT COMPONENTS
import { FormPartsContainer } from './FormPartsContainer';
import { AttentionArrow } from './formHelpers';
import { CashFlowInputsRow } from './cashflow';
import { ControlledRadioYesNo } from './inputs';
import {
  ContentH1,
  IntervalColumnHeadings,
} from '../components/headings';
import {
  ContractRentField,
  RentShareField,
  PlainRentRow,
} from './rentFields';
import { HeadingWithDetail } from '../components/details';
// Premature feature temporarily hidden to avoid messy revert
// import { ExpensesOther } from './ExpensesOther';
import { ShowOnYes } from './ShowOnYes';

// LOGIC
import {
  getEveryMember,
  isHeadOrSpouse,
  getDependentMembers,
  isDisabled,
  isUnder13,
} from '../utils/getMembers';
import { getUnder13Expenses } from '../utils/cashflow';


// ========================================
// COMPONENTS
// ========================================

/** Renders a yes/no choice that will reveal the cash
 *     flow component given when the user selects 'yes'.
 *
 * @note: We added this extra step between the user and
 * the input because people kept skipping that question.
 *
 * @param {object} props
 * @param {bool} props.hasExpenses True if client has any
 *     expenses here that could affect their income.
 * @param {object} props.CashFlowRow To be rendered if user
 *     chooses 'yes'.
 * @param {string | object} props.label To be rendered as
 *     the yes/no question.
 * @param {object} props.propData Data for the prop changed
 *     by the given cash flow component. (move component in
 *     here?)
 * @param {string} props.propData.childPropName Name of cash
 *     flow client prop to be updated.
 * @param {object} props.propData client Current or future
 *     client data.
 * @param {function} props.propData update Updates client
 *     values
 *
 * @returns Value that React can render
 */
const EarnedFrom = function ({ hasExpenses, CashFlowRow, label, propData }) {

  /** @todo Save amount temporarily when 'source'
   *      amount is set to 0. */
  var reset = function (evnt) {
    var { childPropName, update } = propData;

    update(evnt, {
      name:  childPropName,
      value: 0,
    });
  };

  if (hasExpenses) {

    var { childPropName, client } = propData;
    var showProps = {
      childName:           childPropName,
      showChildrenAtStart: client[ childPropName ] > 0,
      question:            label,
      heading:             null,
      onNo:                reset,
      // `<Surrounder>` props
      Left:                <AttentionArrow />,
    };

    return (
      <div className= { 'earned-from' }>
        <ShowOnYes { ...showProps }>
          { CashFlowRow }
        </ShowOnYes>
      </div>
    );

  } else {
    return null;
  }

};  // End EarnedFrom


const Utilities = function ({ current, type, time, updateClientValue }) {

  let hasClimate     = current.climateControl,
      hasElectricity = current.nonHeatElectricity,
      hasPhone       = current.phone,
      hasFuelAssist  = current.fuelAssistance;

  let setChecked = function (evnt, inputProps) {
    var obj = { ...inputProps, value: inputProps.checked };
    updateClientValue(evnt, obj);
  };  // End setChecked()

  // For keyboard access (already does spacebar)
  let onKeyDown = function (evnt) {
    if (evnt.key === `Enter`) {
      evnt.target.click();
    }
  };

  // May want to change name to 'utilities' and value to what's 'name' now
  // Will require more work in the change handler
  return (
    <div>
      <Header as='h4'>Which of these utilities do you pay for?</Header>

      <Checkbox
        name={ 'climateControl' }
        label={ 'Heating or cooling (e.g. A/C during summer)' }
        checked={ hasClimate }
        onChange={ setChecked }
        onKeyDown = { onKeyDown } />
      <br />
      <Checkbox
        name={ 'nonHeatElectricity' }
        label={ 'Electricity for non-heating purposes' }
        checked={ hasElectricity }
        onChange={ setChecked }
        onKeyDown = { onKeyDown } />
      <br />
      <Checkbox
        name={ 'phone' }
        label={ 'Telephone service' }
        checked={ hasPhone }
        onChange={ setChecked }
        onKeyDown = { onKeyDown } />
      <br />
      <br />
      <ControlledRadioYesNo
        labelText = { 'Do you get Fuel Assistance?' }
        checked   = { hasFuelAssist }
        name      = { 'fuelAssistance' }
        onChange  = { updateClientValue } />

    </div>

  );
};  // End Utilities(<>)


const HousingDetails = function ({ current, type, time, updateClientValue }) {

  let housing = current.housing,
      sharedProps = {
        timeState:         current,
        current:           current,
        type:              type,
        time:              time,
        updateClientValue: updateClientValue,
      };

  if (current.housing === 'voucher') {
    return (
      <div>
        <ContractRentField { ...sharedProps } />
        <RentShareField { ...sharedProps } />
        <Utilities { ...sharedProps } />
      </div>
    );

  } else if (housing === 'homeless') {
    return null;

  } else if (housing === 'renter') {
    return (
      <div>
        <br />
        <PlainRentRow { ...sharedProps } />
        <Utilities { ...sharedProps } />
      </div>
    );

  } else if (housing === 'homeowner') {
    return (
      <div>
        <IntervalColumnHeadings type={ type } />
        <CashFlowInputsRow
          { ...sharedProps }
          generic={ 'mortgage' }> Mortgage
        </CashFlowInputsRow>
        <CashFlowInputsRow
          { ...sharedProps }
          generic={ 'housingInsurance' }> Insurance Costs
        </CashFlowInputsRow>
        <CashFlowInputsRow
          { ...sharedProps }
          generic={ 'propertyTax' }> Property Tax
        </CashFlowInputsRow>
        <Utilities { ...sharedProps } />
      </div>
    );

  }  // end which expenses
};  // End HousingDetails(<>)


const HousingRadio = function ({ currentValue, label, time, updateClientValue }) {

  var value = label.toLowerCase();

  return (
    <Form.Field>
      <Radio
        name={ 'housing' }
        label={ label }
        value={ value }
        checked={ currentValue === value }
        onChange = { updateClientValue } />
    </Form.Field>
  );

};  // End HousingRadio(<>)


/**
 * @function
 * @param {object} props
 * @param {object} props.current Client data of current user circumstances
 * @param {string} props.type 'expense' or 'income', etc., for classes
 * @param {string} props.time 'current' or 'future'
 * @param {function} props.updateClientValue Sets state values
 *
 * @returns React element
 */
const Housing = function ({ current, type, time, updateClientValue }) {

  /** @deprecated This is handled differently now */
  let ensureRouteAndValue = function (evnt, inputProps) {
    var obj = { ...inputProps, name: inputProps.name, value: inputProps.value, checked: null };
    updateClientValue(evnt, obj);
  };

  let sharedProps = {
    current:           current,
    type:              type,
    time:              time,
    updateClientValue: ensureRouteAndValue,
  };

  return (
    <div>

      <ContentH1>Housing</ContentH1>

      { current.housing === 'voucher' ? (
        null
      ) : (
        <div>
          <Header as='h4'>What is your housing situation?</Header>
          <HousingRadio
            currentValue={ current.housing }
            label={ 'Homeless' }
            time={ time }
            updateClientValue = { ensureRouteAndValue } />
          <HousingRadio
            currentValue={ current.housing }
            label={ 'Renter' }
            time={ time }
            updateClientValue = { ensureRouteAndValue } />
          <HousingRadio
            currentValue={ current.housing }
            label={ 'Homeowner' }
            time={ time }
            updateClientValue = { ensureRouteAndValue } />
        </div>
      ) }

      <HousingDetails { ...sharedProps } />

    </div>
  );

};  // End Housing()


/** Abstracted for future use in 'future' value setting as well.
 * @function
 * @param {object} props
 * @param {object} props.current Client data of current user circumstances
 * @param {string} props.time 'current' or 'future'
 * @param {function} props.updateClientValue Sets state values
 * @param {object} props.snippets Language-specific content
 *
 * @returns React element
 */
const ExpensesFormContent = function ({ current, time, updateClientValue, snippets }) {

  let type        = 'expense',
      household   = current.household,
      sharedProps = {
        timeState:         current,
        type:              type,
        time:              time,
        updateClientValue: updateClientValue,
      };

  /* @todo Make a more general age-checking function to keep
   * household data format under control in one place. */
  var isOver12 = function (member) {
    return !isUnder13(member);
  };

  // Won't include head or spouse
  var allDependents = getDependentMembers(household),
      under13       = getEveryMember(allDependents, isUnder13),
      over12        = getEveryMember(allDependents, isOver12);

  // 'Elderly' here is using the lowest common denominator - SNAP standards.
  var isElderlyOrDisabled = function (member) {
    return isDisabled(member) || member.m_age >= 60;
  };
  var elderlyOrDisabled = getEveryMember(household, isElderlyOrDisabled),
      elderlyOrDisabledHeadOrSpouse = getEveryMember(elderlyOrDisabled, isHeadOrSpouse);

  return (
    <div className='field-aligner two-column'>

      { under13.length > 0 ? (
        <div>
          <ContentH1 subheading = { snippets.unreimbursedNonMedicalChildCare.i_subheading }>
            { snippets.unreimbursedNonMedicalChildCare.i_sectionHeading }
          </ContentH1>
          <IntervalColumnHeadings type={ type } />
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'childDirectCare' }>
            { snippets.unreimbursedNonMedicalChildCare.childDirectCare.i_label }
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'childBeforeAndAfterSchoolCare' }>
            { snippets.unreimbursedNonMedicalChildCare.childBeforeAndAfterSchoolCare.i_label}
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'childTransportation' }>
            { snippets.unreimbursedNonMedicalChildCare.childTransportation.i_label }
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'childOtherCare' }>
            { snippets.unreimbursedNonMedicalChildCare.childOtherCare.i_label }
          </CashFlowInputsRow>

          <EarnedFrom
            hasExpenses = { getUnder13Expenses(current) !== 0 }
            label    = { `If you didn't have that child care, would it change how much pay you can bring home?` }
            propData = {{
              client:        current,
              childPropName: `earnedBecauseOfChildCare`,
              update:        updateClientValue,
            }}
            CashFlowRow = {
              <CashFlowInputsRow
                { ...sharedProps }
                generic = { `earnedBecauseOfChildCare` }>
                { `How much less would you make?` }
              </CashFlowInputsRow>
            } />

        </div>
      ) : (
        null
      ) }

      { current.hasSnap ? (
        <div>
          <ContentH1>Child Support</ContentH1>
          <IntervalColumnHeadings type={ type } />
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'childSupportPaidOut' }> <strong>Legally obligated</strong> child support
          </CashFlowInputsRow>
        </div>
      ) : (
        null
      ) }

      {/* Head or spouse can't be a dependent, so they don't count. */}
      { over12.length > 0 ? (
        <div>
          <ContentH1 subheading = { 'For the care of people who are older than 12, but are still dependents (those under 18 or disabled). Don\'t include amounts that are paid for by other benefit programs.\n' }>
            Dependent Care of Persons Over 12 Years of Age
          </ContentH1>
          <IntervalColumnHeadings type={ type } />
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'adultDirectCare' }> Direct care costs
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'adultTransportation' }> Transportation costs
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'adultOtherCare' }> Other care
          </CashFlowInputsRow>
        </div>
      ) : (
        null
      ) }

      { elderlyOrDisabled.length > 0 ? (
        <div>
          <HeadingWithDetail>
            <ContentH1>Unreimbursed Disabled/Handicapped/Elderly Assistance</ContentH1>
            <div>
              <div>Unreimbursed expenses to cover care attendants and auxiliary apparatus for any family member who is elderly or is a person with disabilities. Auxiliary apparatus are items such as wheelchairs, ramps, adaptations to vehicles, or special equipment to enable a blind person to read or type, but only if these items are directly related to permitting the disabled person or other family member to work.</div>
              <div>Examples of eligible disability assistance expenses:</div>
              <ul>
                <li>The payments made on a motorized wheelchair for the 42 year old son of the head of household enable the son to leave the house and go to work each day on his own. Prior to the purchase of the motorized wheelchair, the son was unable to make the commute to work. These payments are an eligible disability assistance expense.</li>
                <li>Payments to a care attendant to stay with a disabled 16-year-old child allow the child’s mother to go to work every day. These payments are an eligible disability assistance allowance.</li>
              </ul>
            </div>
          </HeadingWithDetail>
          <IntervalColumnHeadings type={ type } />
          <CashFlowInputsRow
            { ...sharedProps }
            generic={ 'disabledAssistance' }> Disabled/Handicapped assistance
          </CashFlowInputsRow>

          <EarnedFrom
            hasExpenses = { current.disabledAssistance !== 0 }
            label    = { `If you didn't have that assistance, would it change how much pay you can bring home?` }
            propData = {{
              client:        current,
              childPropName: `earnedBecauseOfAdultCare`,
              update:        updateClientValue,
            }}
            CashFlowRow = {
              <CashFlowInputsRow
                { ...sharedProps }
                generic = { `earnedBecauseOfAdultCare` }>
                { `How much less would you make?` }
              </CashFlowInputsRow>
            } />

        </div>
      ) : (
        null
      ) }

      {/* This comment may or may not belong here. Deeper discussion needed:
        * These medical expenses count for Section 8 too if the disabled
        *     person is the head or spouse. From Appendix B, item (D)
        *     {@link http://www.tacinc.org/media/58886/S8MS%20Full%20Book.pdf} */}
      { elderlyOrDisabledHeadOrSpouse.length > 0 || (current.hasSnap && elderlyOrDisabled.length > 0) ? (
        <div>
          <HeadingWithDetail>
            <ContentH1>Unreimbursed Medical Expenses</ContentH1>
            <div>
              <div>Do not repeat anything you already listed in the section above. Examples of allowable medical expenses:</div>
              <ul>
                <li>The orthodontist expenses for a child’s braces.</li>
                <li>Services of doctors and health care professionals.</li>
                <li>Services of health care facilities.</li>
                <li>Medical insurance premiums. </li>
                <li>Prescription/non-prescription medicines (prescribed by a physician).</li>
                <li>Transportation to treatment (cab fare, bus fare, mileage).</li>
                <li>Dental expenses, eyeglasses, hearing aids, batteries.</li>
                <li>Live-in or periodic medical assistance.</li>
                <li>Monthly payment on accumulated medical bills (regular monthly payments on a bill that was previously incurred).</li>
              </ul>
            </div>
          </HeadingWithDetail>
          <IntervalColumnHeadings type={ type } />
          <CashFlowInputsRow
            { ...sharedProps }
            generic='disabledMedical'> Disabled/Elderly medical expenses
          </CashFlowInputsRow>
          <CashFlowInputsRow
            { ...sharedProps }
            generic='otherMedical'> Medical expenses of other members
          </CashFlowInputsRow>
        </div>
      ) : (
        null
      ) }

      <Housing
        current={ current }
        time={ time }
        type={ type }
        updateClientValue = { updateClientValue } />

      {/* Premature feature temporarily hidden to avoid messy revert
        <ShowOnYes
          clientPartial = { current }
          propName = { `wantsToSeeOtherExpenses` }
          updateClientValue = { updateClientValue }
          question = { `Do you want to enter your other expenses so you can see if you need to make a different plan?` }
          heading = { `Other Expenses` }>
          <ExpensesOther { ...sharedProps } />
        </ShowOnYes>
      */}
    </div>
  );

};  // End ExpensesFormContent()

/* Move these to SNAP calc logic scripts:
 * @todo SNAP: Does a medical assistant's payments count as a medical expense?
 *     (Answer: Yes. @see {@link https://www.mass.gov/service-details/snap-verifications})
 * @todo SNAP: Medical expense only matters if household has elder/disabled, but
 *     are they any medical expenses or only those of the disabled person? "Medical
 *     Expenses for Disabled or Elderly". Also, do they sometimes need to
 *     enter medical expenses even if they don't have an elderly or disabled
 *     household member?
 */

// `props` is a cloned version of the original props. References broken.
/**
  * @function
  * @param {object} props
  * @param {function} props.updateClientValue Setting client state
  * @param {object} props.navData Buttons for bottom row
  * @param {object} props.client Data for calculating benefits, `future` and `current`
  * @param {object} props.snippets  Language-specific content
  *
  * @returns React element
  */
const CurrentExpensesStep = function ({ updateClientValue, navData, client, snippets }) {

  return (
    <FormPartsContainer
      title     = { snippets.i_title }
      clarifier = { snippets.i_clarifier }
      navData   = { navData }
      formClass = { `expenses` }>
      <ExpensesFormContent
        updateClientValue = { updateClientValue }
        current={ client.current }
        time={ 'current' }
        snippets={ snippets } />
    </FormPartsContainer>
  );

};  // End CurrentExpensesStep()


export { CurrentExpensesStep };
