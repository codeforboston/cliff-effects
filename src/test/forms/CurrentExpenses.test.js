import React from 'react';
import { mount } from 'enzyme';

import { CurrentExpensesStep } from '../../forms/CurrentExpenses';

import { CLIENT_DEFAULTS } from '../../utils/CLIENT_DEFAULTS';

// LOCALIZATION
import { getTextForLanguage } from '../../utils/getTextForLanguage';


var snippets = getTextForLanguage(`en`);
test('Expenses step component should render without error', () => {
  const navData = {
          left:   (<div>Exp left</div>),
          middle: null,
          right:  (<div>Exp right</div>),
        },
        updateClientValue = jest.fn(),
        saveForm          = jest.fn(),
        askToResetClient  = jest.fn(),
        openFeedback      = jest.fn(),
        formSnippets      = snippets.visitPage.currentExpenses;

  expect(() => {
    mount(
      <CurrentExpensesStep
        client            = { CLIENT_DEFAULTS }
        navData           = { navData }
        updateClientValue = { updateClientValue }
        saveForm          = { saveForm }
        askToResetClient  = { askToResetClient }
        openFeedback      = { openFeedback }
        snippets          = { formSnippets } />
    );
  }).not.toThrow();
});
