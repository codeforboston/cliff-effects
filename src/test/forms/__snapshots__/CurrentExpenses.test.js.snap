// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expenses step component renders as snapshot correctly 1`] = `
<Form
  as="form"
  className="expense-form flex-item flex-column"
>
  <FormPartsContainer
    navData={
      Object {
        "left": Object {
          "onClick": [Function],
          "text": "Previous",
        },
        "middle": null,
        "right": Object {
          "onClick": [Function],
          "text": "Next",
        },
      }
    }
    title={
      <span
        className="snippet"
        lang="en"
      >
        Current Household Expenses
      </span>
    }
  >
    <ExpensesFormContent
      current={
        Object {
          "SSDI": 0,
          "SSI": 0,
          "TAFDC": 0,
          "adultDirectCare": 0,
          "adultOtherCare": 0,
          "adultTransportation": 0,
          "alimony": 0,
          "childBeforeAndAfterSchoolCare": 0,
          "childDirectCare": 0,
          "childOtherCare": 0,
          "childSupportIn": 0,
          "childSupportPaidOut": 0,
          "childTransportation": 0,
          "climateControl": false,
          "contractRent": 0,
          "disabledAssistance": 0,
          "disabledMedical": 0,
          "earned": 0,
          "earnedBecauseOfAdultCare": 0,
          "earnedBecauseOfChildCare": 0,
          "fuelAssistance": false,
          "hasSection8": false,
          "hasSnap": false,
          "household": Array [
            Object {
              "m_age": 30,
              "m_disabled": false,
              "m_role": "head",
            },
          ],
          "housing": "homeless",
          "housingInsurance": 0,
          "mortgage": 0,
          "nonHeatElectricity": false,
          "otherExpensesCable": 0,
          "otherExpensesCareProducts": 0,
          "otherExpensesClothes": 0,
          "otherExpensesEntertainment": 0,
          "otherExpensesFood": 0,
          "otherExpensesMedical": 0,
          "otherExpensesOther": 0,
          "otherExpensesPhone": 0,
          "otherExpensesTransport": 0,
          "otherExpensesUtilities": 0,
          "otherIncome": 0,
          "otherMedical": 0,
          "pension": 0,
          "phone": false,
          "propertyTax": 0,
          "rent": 0,
          "rentShare": 0,
          "socialSecurity": 0,
          "unemployment": 0,
          "wantsToSeeOtherExpenses": false,
          "workersComp": 0,
        }
      }
      snippets={
        Object {
          "childSupport": Object {
            "childSupportPaidOut": Object {
              "i_childSupport": <span
                className="snippet"
                lang="en"
        >
                 child support
        </span>,
              "i_legallyObligated": <span
                className="snippet"
                lang="en"
        >
                Legally obligated
        </span>,
              "langCode": "en",
            },
            "i_columnExpenseType": <span
              className="snippet"
              lang="en"
        >
              Expense
        </span>,
            "i_sectionHeading": <span
              className="snippet"
              lang="en"
        >
              Child Support
        </span>,
            "langCode": "en",
          },
          "housing": Object {
            "climateControl": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay if you have a separate bill for heating and/or cooling
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Heating or cooling (e.g. A/C during summer)
        </span>,
              "langCode": "en",
            },
            "fuelAssistance": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                Fuel Assistance helps you pay for heating fuel
        </span>,
              "i_labelText": <span
                className="snippet"
                lang="en"
        >
                Do you get Fuel Assistance?
        </span>,
              "langCode": "en",
            },
            "i_sectionHeading": <span
              className="snippet"
              lang="en"
        >
              Housing
        </span>,
            "i_utilitiesSubheading": <span
              className="snippet"
              lang="en"
        >
              Which of these utilities do you pay for?
        </span>,
            "langCode": "en",
            "monthlyContractRent": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                The total rent for your apartment
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Monthly Contract Rent (the total rent for your apartment)
        </span>,
              "langCode": "en",
            },
            "monthlyRentShare": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much of the total rent you have to pay
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Your Monthly Rent Share (how much of the total rent you have to pay)
        </span>,
              "langCode": "en",
            },
            "nonHeatElectricity": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for any electric usage (other than for heat)
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Electricity for non-heating purposes
        </span>,
              "langCode": "en",
            },
            "phone": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for basic telephone service
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Telephone service
        </span>,
              "langCode": "en",
            },
          },
          "i_next": <span
            className="snippet"
            lang="en"
        >
            Next
        </span>,
          "i_previous": <span
            className="snippet"
            lang="en"
        >
            Previous
        </span>,
          "i_title": <span
            className="snippet"
            lang="en"
        >
            Current Household Expenses
        </span>,
          "langCode": "en",
          "unreimbursedNonMedicalChildCare": Object {
            "childBeforeAndAfterSchoolCare": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for child care for times before or after school?
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Before- and after-school care
        </span>,
              "langCode": "en",
            },
            "childDirectCare": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for child care out of pocket?
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Direct care costs
        </span>,
              "langCode": "en",
            },
            "childOtherCare": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for other child care?
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Other care
        </span>,
              "langCode": "en",
            },
            "childTransportation": Object {
              "i_hint": <span
                className="snippet"
                lang="en"
        >
                How much do you pay for transportation?
        </span>,
              "i_label": <span
                className="snippet"
                lang="en"
        >
                Transportation costs
        </span>,
              "langCode": "en",
            },
            "i_columnExpenseType": <span
              className="snippet"
              lang="en"
        >
              Expense
        </span>,
            "i_doEarnBecauseOfChildCare": <span
              className="snippet"
              lang="en"
        >
              Does child care allow you to make additional income?
        </span>,
            "i_earnedBecauseOfChildCare": <span
              className="snippet"
              lang="en"
        >
              Income made possible by childcare expenses
        </span>,
            "i_sectionHeading": <span
              className="snippet"
              lang="en"
        >
              Reasonable Unreimbursed Non-Medical Child(ren) Care
        </span>,
            "i_subheading": <span
              className="snippet"
              lang="en"
        >
              A "child" is a person 12 or younger. Don't include amounts that are paid for by other benefit programs.
        </span>,
            "langCode": "en",
          },
        }
      }
      time="current"
      updateClientValue={[Function]}
    />
  </FormPartsContainer>
</Form>
`;
